### PR TITLE
Add file name lint go script

### LINKED ISSUE #163: Rewrite shell scripts in Go
I think we could try to rewrite our [Bash-based linting scripts](https://github.com/pion/.goassets/tree/master/scripts) in Go.

WDYT?

### PR DESCRIPTION
**Pull request by kehiy**
#### Description
rewrite the lint file name script in Golang.

#### Reference issue
- Related #163

### CODE REVIEW

**Comment by kehiy** [comment]
@Sean-Der can you have a review?

File: scripts/lint_filename.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
+SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+SPDX-License-Identifier: MIT
+*/
```
Comment (stv0g) [inline scripts/lint_filename.go:4]:
```suggestion
// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
// SPDX-License-Identifier: MIT
```

File: scripts/lint_filename.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	"os"
+	"path/filepath"
+	"regexp"
+)
+
+func main() {
+	parentDir := filepath.Dir(os.Args[0])
+	goRegex := regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_]*\.go$`)
+
+	err := filepath.Walk(parentDir, func(path string, info os.FileInfo, err error) error {
```
Comment (stv0g) [inline scripts/lint_filename.go:18]:
```suggestion
	if err := filepath.Walk(parentDir, func(path string, info os.FileInfo, err error) error {
```

File: scripts/lint_filename.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+		filename := filepath.Base(path)
+
+		if !goRegex.MatchString(filename) {
+			return fmt.Errorf("%s is not a valid filename for Go code, only alpha, numbers and underscores are supported", filename)
+		}
+
+		return nil
+	})
+	if err != nil {
```
Comment (stv0g) [inline scripts/lint_filename.go:39]:
```suggestion
	}); err != nil {
```

I usually prefer this shorter form. But its a personal preference. Feel free to accept or decline this suggestion how you like :)

**Review by stv0g — CHANGES_REQUESTED** [review]
Looks good to me :) We now only need to invoke the new Go script instead of the shell script here:

https://github.com/pion/.goassets/blob/965ecf8cc04ca3535ee99556c40b3cc27af671d2/.github/workflows/lint.reusable.yml#L35

**Comment by kehiy** [comment]
@stv0g is that okay now?

**Comment by stv0g** [comment]
@kehiy It seems like you have resolved my previous review comments without addressing them?

Please also not that the CI is currently failing. We can only merge it once all the checks are passing.

Thanks for your support :)

File: .github/workflows/lint.reusable.yml [hunk 4]
```diff
@@ -32,7 +33,7 @@ jobs:
         run: .github/.goassets/scripts/lint-commit-message.sh
 
       - name: File names
-        run: .github/.goassets/scripts/lint-filename.sh
+        run: .github/.goassets/scripts/lint-filename.go
```
Comment (stv0g) [inline .github/workflows/lint.reusable.yml:36]: I assume we need to use a `go run .github/.goassets/scripts/lint-filename.go` here.
```suggestion
        run: go run .github/.goassets/scripts/lint-filename.go
```

**Comment by kehiy** [comment]
@stv0g what so you mean `addressing them`?
about ci I tried to fix them but its true we can merge it

**Comment by stv0g** [comment]
> @stv0g what so you mean addressing them?

I've proposed changes which you have ignored and marked the corresponding threads as resolved.
I have now went ahead an unresolved them. Please commit the suggested changes.

**Comment by kehiy** [comment]
> > @stv0g what so you mean addressing them?
> 
> I've proposed changes which you have ignored and marked the corresponding threads as resolved. I have now went ahead an unresolved them. Please commit the suggested changes.

oh, yeas I can see, but I committed the suggestions on GitHub directly!
I will try to do it againg.

**Comment by kehiy** [comment]
@stv0g I think its ok now
Is that?

**Comment by stv0g** [comment]
Okay cool :) It looks alright now 👍🏻 

You would only need to fix the CI pipeline. Here are some tips:

1. Add the same SPDX headers to the `go.mod` mod file in the beginning
2. Add your name + email (`Kay <kehiiiiya@gmail.com>`) to the `AUTHORS.txt` file. Make sure the entry is at the correct place (alphabetically ordered)
3. [Squash all your commits](https://www.baeldung.com/ops/git-squash-commits) and give it a correct commit message.

**Comment by kehiy** [comment]
@stv0g Everything is fixed just for `squashing` the commits I got a lot of errors and conflicts, and I think it's better to squash them when you want to merge them on GitHub.

**Comment by stv0g** [comment]
Hi, you can also squash lically. I added a link about it in my previous comment

**Comment by kehiy** [comment]
okay, I tried to do this locally, but I will try again.

**Comment by kehiy** [comment]
@stv0g can you have a review?

**Review by Sean-Der — APPROVED** [review]

**Pull request merged**